validator/src/journal/block_manager.rs

```diff
@@ -398,13 +402,158 @@ impl BlockManagerState {
 #[derive(Default, Clone)]
 pub struct BlockManager {
     state: Arc<BlockManagerState>,
+    persist_lock: Arc<RwLock<()>>,
 }
 
 impl BlockManager {
     pub fn new() -> Self {
         BlockManager::default()
     }
 
+    fn transaction_index_contains(
+        &self,
+        index: &Box<IndexedBlockStore>,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let index = index.as_ref();
+        if TransactionIndex::contains(index, id)? {
+            let block = TransactionIndex::get_block_by_id(index, id)?
+                .expect("BlockStore updated during transaction index check");
```

Comment:
These expects are a concern, as they imply that it is possible for the store to be updated concurrently in-between calls. You can eliminate this one by eliminating the call to contains() and just using get_block_by_id() directly. The second one requires holding the lock on `block_by_block_id` during the duration of this call to ensure it, and the underlying blockstores, are not modified while looking up the transaction.

Reply:
Yeah, this could be clearer. `if let Some(block) = TransactionIndex::get_block_by_id(index, id)` is how I should go.